managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java

```diff
@@ -378,63 +378,77 @@ public void asyncOpen(final String name, final ManagedLedgerConfig config, final
+                        // May need to trigger offloading
+                        if (config.isTriggerOffloadOnTopicLoad()) {
+                            newledger.maybeOffloadInBackground(NULL_OFFLOAD_PROMISE);
+                        }
                     }
 
-                    // Clean the map if initialization fails
-                    ledgers.remove(name, future);
+                    @Override
+                    public void initializeFailed(ManagedLedgerException e) {
+                        if (config.isCreateIfMissing()) {
+                            log.error("[{}] Failed to initialize managed ledger: {}", name, e.getMessage());
+                        }
 
-                    if (pendingInitializeLedgers.remove(name, pendingLedger)) {
-                        pendingLedger.ledger.asyncClose(new CloseCallback() {
-                            @Override
-                            public void closeComplete(Object ctx) {
-                                // no-op
-                            }
+                        // Clean the map if initialization fails
+                        ledgers.remove(name, future);
 
-                            @Override
-                            public void closeFailed(ManagedLedgerException exception, Object ctx) {
-                                log.warn("[{}] Failed to a pending initialization managed ledger", name, exception);
-                            }
-                        }, null);
-                    }
+                        if (pendingInitializeLedgers.remove(name, pendingLedger)) {
+                            pendingLedger.ledger.asyncClose(new CloseCallback() {
+                                @Override
+                                public void closeComplete(Object ctx) {
+                                    // no-op
+                                }
 
-                    future.completeExceptionally(e);
-                }
-            }, null);
+                                @Override
+                                public void closeFailed(ManagedLedgerException exception, Object ctx) {
+                                    log.warn("[{}] Failed to a pending initialization managed ledger", name, exception);
+                                }
+                            }, null);
+                        }
+                        future.completeExceptionally(e);
+                    }
+                }, null);
+                return null;
+            });
             return future;
         }).thenAccept(ml -> callback.openLedgerComplete(ml, ctx)).exceptionally(exception -> {
             callback.openLedgerFailed((ManagedLedgerException) exception.getCause(), ctx);
             return null;
         });
     }
 
+    private CompletableFuture<BookKeeper> createBookKeeperClient(ManagedLedgerConfig config) {
+        CompletableFuture<BookKeeper> future = new CompletableFuture<>();
+        scheduledExecutor.execute(() -> {
```

Comment:
I think the culprit of the issue is in the `BookkeeperFactoryForCustomEnsemblePlacementPolicy` interface, because it effectively masks a potentially blocking operation behind a method that does not look blocking, and doesn't even declare an exception.

I think a better solution here would to make the interface `BookkeeperFactoryForCustomEnsemblePlacementPolicy` to return a CompletableFuture of a BK client instead of using an executor here.

Reply:
> I think the culprit of the issue is in the BookkeeperFactoryForCustomEnsemblePlacementPolicy interface

Yes. But that's something that depends on other system which is BookKeeper in this case, Upgrading bk version will not be possible for all Pulsar stable version. 
Let's merge this patch and I can make bookkeeper enhancement to fix this issue for future version.